### DISCUSSION TITLE
Multipage apps (MPA), sessions, and auth with JSVM?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#5768] [Q&A] [answered]
Gani, following up on our [brief discussion](https://github.com/pocketbase/pocketbase/discussions/5307#discussioncomment-10942216) about loading auth state from a cookie instead of JWT.

I've finally reached a point where I'm ready to begin exploring a cookie-based server-side login/auth strategy. As you might remember, I'm working on pocketpages.dev which uses the JSVM to do server-side pages and multi-page apps (MPA). It's getting a lot of interest from users.

I'd like to add a basic cookie session and auth support:

```html
<form method="post">
  <label>Login</label>
  <input name="login" />
  <label>Password</label>
  <input name="password" type="password" />
  <button type="submit">Login</button>
</form>
```

As I understand it, PB works only with JWT right now. Here are some questions:

1. Once I have the POST data, is there a JSVM equivalent to `authWithPassword("test@example.com", "1234567890")` that could give back a token that I could return as a cookie?
2. On subsequent requests, is there any way to write some JSVM middleware that looks for the auth cookie and either performs the equivalent functions or converts it to a JWT that can be passed along? Something equivalent to `c.Set(ContextAdminKey, admin)` in v22 or assigning `e.Auth` in v23?

If this is not currently possible, I would be interested in helping make a PR for it if you are open.

As an alternative, it looks like libraries like htmx can do custom request headers, so it would be possible to get the JWT from the PocketBase JS client and forward it. That is a partial solution, but full page loads would not benefit from it:

```js
document.body.addEventListener('htmx:configRequest', function(event) {
  var xhr = event.detail.xhr;
  xhr.setRequestHeader('Your-Custom-Header', 'YourHeaderValue');
});
```

Unfortunately, full page loads would still be a problem. I could partially mitigate this by using htmx islands that make a fetch to load their initial data, but that seems suboptimal:

```html
<div hx-get="/your/api/endpoint" hx-trigger="load"></div>
```

Please let me know your thoughts :)

**Comment by benallfree** [comment]
Update: I have managed to obtain a JWT by calling `$http.fetch( { url: '/api/collections/users/auth-with-password' } )`. I'm not sure if there is a more direct way to do it, but this works. I can also use `Set-Cookie` successfully.

Now all that remains is to somehow get the JWT cookie back into the auth middleware.

**Comment by ganigeorgiev** [comment]
@benallfree I'm honestly confused and not sure that I understand your questions but you could find some outdated examples how to implemented a cookie based flow in https://github.com/pocketbase/pocketbase/discussions/989, https://github.com/pocketbase/pocketbase/discussions/3447 and probably others.

But as mentioned in previous discussions, I'm not interested working on SSR helpers for the time being.

Just note that:

1. You don't need to create a PR because it is not required to be part of the main repo for this to work (unless you are planning to change the Admin UI).

2. The JWT is just something trusted based on which we can find the related user. You don't have to use JWT for your custom routes if you don't want to.

3. If you are modifying the default JSON endpoints make sure to add check for admins/superusers because the Admin UI is a static client-side application and relies on them.

4. You'll notice that in both v0.22.x and v0.23.0-rc the middleware that loads the default auth state **doesn't return an error** if no token or user is found, they just leave the request auth state empty. This allow users to register their own external middlewares and load the auth state on their own. If you want a request to be considered authenticated you need to assign an auth record instance into the request context (or `e.Auth` in v0.23) and every following middleware that checks for the auth state just reads the value from the request event/context (_in addition with v0.23 there is also support for specifying priority of the execution and replace/unbind of existing middlewares based on their id_).

**Comment by benallfree** [comment]
It's okay thanks I got it :) 

I ported your JWT middleware to JSVM and made it look for a cookie as well. That did it.

Final question: is there a JSVM function like `authWithPassword` from the JS SDK? Something so I don't need to call the API on itself?

**Reply by ganigeorgiev** [reply]
Please try to be more specific. Again, I don't understand what do you mean by "JSVM function like `authWithPassword`". What does this even mean? What is the expected behavior you want? Did you checked the linked discussions above and more specifically https://github.com/pocketbase/pocketbase/discussions/3447#discussioncomment-7201367 and https://github.com/pocketbase/pocketbase/discussions/3447#discussioncomment-7605828?

**Reply by benallfree** [reply]
The JS client SDK has a method:

```js
const userData = await pb.collection('users').authWithPassword('test@example.com', '123456');
```

I understand that hits `/api/collections/users/auth-with-password` which returns a JWT token and user Record upon success.

Is there something in the [JSVM](https://pocketbase.io/jsvm) that accomplishes the same thing? The expected behavior is something like:

```js
// auth.pb.js
const { token, record } = $app.authWithPassword('test@example.com', '123456')
```

**Reply by ganigeorgiev** [reply, marked as answer]
So you want to validate a password and generate an auth token?

If that's what you mean then No, there is no direct helper for this. You need to:
- retrieve the record (e.g. `$app.dao().findAuthRecordByEmail("users", "test@example.com")`)
- validate its password (e.g. `record.validatePassword("123456")`)
- generate auth token using the [$tokens helper](https://pocketbase.io/jsvm/modules/_tokens.html) (e.g. `const token = $tokens.recordAuthToken($app, record)`)

**Comment by benallfree** [comment]
In case anyone is curious, here is a snippet from the code I implemented. Keep in mind that this uses some helpers and wrappers, but the essence is there:

```js
const authWithPassword = (email, password) => {
        try {
          const record = $app.dao().findAuthRecordByEmail('users', email)
          if (!record.validatePassword(password)) {
            return null
          }
          const token = $tokens.recordAuthToken($app, record)
          return { token, user: JSON.parse(stringify(record)) }
        } catch (e) {
          dbg(e)
          return null
        }
      }
```

**Reply by ganigeorgiev** [reply]
Side-note: If you just want a plain map with the public safe fields of the record, a better alternative of `JSON.parse(stringify(record))` could be `record.publicExport()`.

**Reply by benallfree** [reply]
`record.publicExport()` seems to convert JSON fields to an array of numbers.

**Reply by ganigeorgiev** [reply]
Could you give an example? What `stringify` does in the above?

`publicExport` essentially creates a map with `"fieldName": record.get("fieldName")` key-value entries.

**Reply by ganigeorgiev** [reply]
Oh, I think I know what do you mean.

The raw field value of a `json` type field is `types.JSONRaw` which is a slice of bytes, so Yes in this case a field value could be an array of ints.

**Comment by mkroemer** [comment]
How is the process of authenticate in the backend with 0.23? The auth-with-password endpoint doesn't exist anymore and i need the auth token of a specific user for requests.

**Reply by ganigeorgiev** [reply]
>  The auth-with-password endpoint doesn't exist anymore

That's not true. Please check https://pocketbase.io/docs/api-records/#auth-with-password.

For all supported auth methods, please read https://pocketbase.io/docs/authentication/.

In any case, my recommendations from above remains and I'd recommend against using PocketBase with htmx because we simply don't have the necessary helpers for such cases and you'll have to build a lot of things on your own, which kind of defeat the purpose of using PocketBase in the first place for me.

**Comment by benallfree** [comment]
htmx and MPA support is now built into https://pocketpages.dev
